### DISCUSSION TITLE
AutoHotkey 1.1.37.02 - PixelCount

### DISCUSSION THREAD

**Discussion by wroma74** [Spawnova/ShinsImageScanClass#29] [General]
Hi Spawnova :-) I am relatively new to coding... I am on Windows 11 Pro, with AutoHotkey 1.1.37.02.

I'd like to use your class to check a pixel in a window outside the desktop and running Chrome, which is not visible and not minimized.

At the moment I am not able to create some code for this purpose, because before I need to learn to make your class runs.

So, in a visible window running Paint, I am experimenting this simple code:

#singleinstance,force
#persistent
#include ShinsImageScanClass.ahk
scan := New ShinsImageScanClass("Senza titolo - Paint")
msgbox % scan.PixelCount(0xED1C24, 3)
WinActivate "Senza titolo - Paint"
CoordMode, Pixel, Client
PixelSearch, Px, Py, 200, 500, 1000, 1000, 0xED1C24, 3, Fast RGB
Sleep, 100
MsgBox, A color within 3 shades of variation was found at X%Px% Y%Py%.
exitapp

As a result, the first msgbox shows me 0, while the second one gives me some coordinates.
It seems your class can not intercept the pixel, while standard PixelSearch can.

Where am I wrong?

Thank you in advance :-)

**Comment by Spawnova** [comment]
Hmm, I'm not entirely sure I tested on my end and it works as expected, one thing you can try is to save the pixel data to an image and inspect it to verify that the script is able to see the correct data
```ahk
#singleinstance,force
#persistent
#include ShinsImageScanClass.ahk
scan := New ShinsImageScanClass("Senza titolo - Paint")
scan.SaveImage("test.png") ;saves to an image exactly what the script sees when scanning
exitapp
```
Another note is that this class uses BitBlt to copy the pixel data from windows, and this is not compatible with all applications, specifically it will likely not be compatible with chrome unless chrome has hardware acceleration disabled

Also to convert the pixelsearch to my class would look like this
```ahk
;PixelSearch, Px, Py, 200, 500, 1000, 1000, 0xED1C24, 3, Fast RGB
if (scan.PixelRegion(0xED1C24,200,500,800,500,3,px,py)) {
    msgbox % "Pixel was found at: " px ", " py
}
```

**Reply by wroma74** [reply]
Thank you, Spawnova!

You were right, if I use your class with MsPaint I obtain black pictures...
Luckily, I have to work on Chrome, and disabling hardware acceleration... it works quite fine, and I obtain nice pixel data in images...

I am working on this automation with Autohotkey v2 and Windows 11:

```autohotkey
#singleinstance
#include ShinsImageScanClass.ahk ;include la library
DetectHiddenWindows(1)
DetectHiddenText(1)
CoordMode "Pixel", "Client"
CoordMode "Mouse", "Client"

foundPix := 0 ; it becomes 1 when a button appears
px:=0 ; x where I look for the button
py:=0 ; y where I look for the button
Windowtitle := "My Hub - Google Chrome"
tmpcount:=0 ;for testing purpose
tmpname:=" " ;for testing purpose


OnExit ExitFunc
scan := ShinsImageScanClass(Windowtitle)
;WinMaximize Windowtitle
;WinActivate Windowtitle
Sleep 333
Loop
{
    timerefresh := Random(1, 3)*1000 ; simulate refresh with various times to make the process "human".
    ControlSend "{F5}",, Windowtitle
    Sleep timerefresh
    foundPix:=scan.PixelRegion(0xF5F5F5,130,298,20,20,0,&px,&py)
    tmpname:="test" String(tmpcount) "F" string(foundPix) ".png"
    scan.SaveImage(tmpname)
    ;msgbox foundPix
    Sleep 100
}
Until foundPix = 1
WinMaximize Windowtitle
WinActivate Windowtitle
Sleep 333
Click 147, 300, "Down"
Sleep 10
Click 147, 300, "Up"
Sleep 10
SoundPlay "C:\Manutenzione\sndtask.mp3"
MsgBox "OK.","YES",64+4096
exitapp

ExitFunc(ExitReason, ExitCode)
{
    if ExitReason != "Exit"
    {
       WinMove 0, 0,,,Windowtitle
       WinMaximize Windowtitle
    }
}
``` 
I am testing the results, and still I have three questions to solve.

Basically, my automation should refresh Chrome and, if a button appears on a specific page, it should sound an alarm, open Chrome and press the button. Meanwhile, I'd like to use my laptop on other tasks and programs, keeping hidden the window with Chrome.

1) Test: If I take a look to Chrome (and I notice there is the button I check for) , then I change window and execute this automation, it works fine and I obtain the following image:
![test13F1](https://github.com/user-attachments/assets/1d7bec6d-58b5-4e2b-8565-ff13b765f06b)

but if I take a look to Chrome (and I notice there is the button I check for), then I change windows lot of times (Excel, Word, MsPaint, Firefox) and then I execute this automation, sadly it does not work and I obtain the following image:
![test5F0](https://github.com/user-attachments/assets/9f220478-e91c-4bee-b043-1e42e55322e2)
Any idea?


2) I need to refresh the Chrome page even if its window is hidden. I added the following code and still have to test it, but do you think it could work?
```autohotkey
    ControlSend "{F5}",, Windowtitle
``` 


3) When will you get tired of all these questions? :-D

Thank you for all you can do, for your patience and for your time :-)

**Reply by Spawnova** [reply]
Controlsend could work yes, you can easily test it by having another window active and trying it out, if the webpage reloads it should be fine.
As to why the image fails, one important thing to keep in mind is that the pixel data can only be obtained if the window is not minimized, I think it can be outside the desktop region but I'm not 100% sure on that, if that is the case you could use `WinMove` to move chrome outside your monitor view until it's found, then `WinMove` it back.

Also, you could look into a chrome extension called tampermonkey, which lets you inject custom javascript on specific websites, you could then have it run a timer checking for that button and make chrome notify and flash the taskbar when it's found, it can also handle refreshing the page. There is even a way to have it run an executable, although it takes some prep work before hand to register it.

If that sounds like something you are interested in you could also ask chatGPT or equivalent for the code via javascript and it shouldn't be too hard to figure out.


If you are still having issues with it, please message me on discord @   .Spawnova   

If you don't want to use discord that is fine as well, but I respond a bit slower on github =P

**Comment by wroma74** [comment]
Hi Spawnova, I want to thank you for your class.

It is really flexible and well done! A little great masterpiece, and I can't wait to see the new version with custom library.

In the end I used AHK 2 and I succeeded in my intent, and while I was writing the code I had many new ideas on what I could do. Thanks again.

I ask you for one more suggestion: I am using your class to monitor an inactive window.
I define

scan := ShinsImageScanClass(WindowsTitle)
scan.UseControlClick := 1

Then, I use scan.Click(147,300,"left")  when a particular event happens in the inactive window.

The focus goes to the inactive window, which then... becomes active.
I would like to avoid this behavior, and leave the window inactive while I continue to work on the active one.

Do you have any ideas?

Thanks for everything.

**Reply by Spawnova** [reply]
ControlClick may not work for all applications, AutoHotkey is not activating the window, but instead when the window receives the mouse event it seems to be activating itself

The implementation in my class is quite basic, you might be able to google another method for performing clicks to inactive windows, sometimes choosing a different control can help or calling just specific windows messages, but I can't say for sure, it's not really something I deal with very often unfortunately =[

**Reply by wroma74** [reply]
Thank you Spawnova, I will try :-)
